llvm.sh

```diff
@@ -94,13 +88,54 @@ case ${DISTRO} in
         ;;
 esac
 
+while getopts ":hm:n:" arg; do
+    case $arg in
+    h)
+        usage
+        ;;
+    m)
+        BASE_URL=${OPTARG}
+        ;;
+    n)
+        CODENAME=${OPTARG}
+        LINKNAME=-${CODENAME}
+        CODENAME_FROM_ARGUMENTS="true"
+        ;;
+    esac
+done
+
+if [[ $EUID -ne 0 ]]; then
+   echo "This script must be run as root!"
+   exit 1
+fi
+
+declare -A LLVM_VERSION_PATTERNS
+LLVM_VERSION_PATTERNS[9]="-9"
+LLVM_VERSION_PATTERNS[10]="-10"
+LLVM_VERSION_PATTERNS[11]="-11"
+LLVM_VERSION_PATTERNS[12]="-12"
+LLVM_VERSION_PATTERNS[13]="-13"
+LLVM_VERSION_PATTERNS[14]="-14"
+LLVM_VERSION_PATTERNS[15]=""
+
+if [ ! ${LLVM_VERSION_PATTERNS[$LLVM_VERSION]+_} ]; then
+    echo "This script does not support LLVM version $LLVM_VERSION"
+    exit 3
+fi
+
+LLVM_VERSION_STRING=${LLVM_VERSION_PATTERNS[$LLVM_VERSION]}
+
 # join the repository name
 if [[ -n "${CODENAME}" ]]; then
-    REPO_NAME="deb http://apt.llvm.org/${CODENAME}/  llvm-toolchain${LINKNAME}${LLVM_VERSION_STRING} main"
+    REPO_NAME="deb ${BASE_URL}/${CODENAME}/  llvm-toolchain${LINKNAME}${LLVM_VERSION_STRING} main"
 
     # check if the repository exists for the distro and version
-    if ! wget -q --method=HEAD http://apt.llvm.org/${CODENAME} &> /dev/null; then
-        echo "Distribution '${DISTRO}' in version '${VERSION}' is not supported by this script."
+    if ! wget -q --method=HEAD ${BASE_URL}/${CODENAME} &> /dev/null; then
+        if [[ -z "${CODENAME_FROM_ARGUMENTS}" ]]; then
```

Comment:
`-z "${CODENAME_FROM_ARGUMENTS}"` is quite error-prone, developers could also assign this as "false" somewhere else accidentally and this evaluates to true

Reply:
It should be safer to set it to reflect user arg `CODENAME_FROM_ARGUMENTS=${OPTARG}` in `-n` option, then have the following changes here:
```suggestion
        if [[ -n "${CODENAME_FROM_ARGUMENTS}" ]]; then
            echo "Specified codename '${CODENAME_FROM_ARGUMENTS}' is not supported by this script."
        else
            echo "Distribution '${DISTRO}' in version '${VERSION}' is not supported by this script."
        fi
```